### ISSUE TITLE
Carp::Heavy - Bizarre copy of HASH in aassign

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2016]
Migrated from [rt.perl.org#3291](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3291) (status was 'resolved')

Searchable as RT3291$

**Comment by p5pRT** [comment]
### From jhoward@fastmail.fm

#### Created by jhoward@fastmail.fm



The following causes a 'Bizarre copy of HASH in aassign' error
perl \-MCGI&#8203;::Carp \-d \-e 'print croak\(1\)'
\<Then press 'c' to continue in the debugger>

It only occurs in debug mode\.

To fix it\, do the following&#8203;:
In Carp&#8203;::Heavy\, change 'DB' to 'DB\_tmp'&#8203;:
Line 39&#8203;: while \(do \{ \{ package DB\_tmp; @&#8203;a = caller\($i\+\+\) \} \} \) \{
Line 79&#8203;: @&#8203;a = @&#8203;\{DB\_tmp&#8203;::args\}; \# must get local copy of args

I don't know if there's any reason the DB package was used in
Carp&#8203;::Heavy\.\.\. If so\, this fix may break something\.

<details><summary>Perl Info</summary>

```
Flags:
    category=library
    severity=high

Site configuration information for perl v5.6.0:

Configured by root at Fri May 19 12:49:37 EST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.12-20, archname=i686-linux
    uname='linux jlinux.optimal-decisions.com 2.2.12-20 #1 thu jan 6
    13:09:01 est 2000 i686 unknown '
    config_args='-de'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef
    usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usesocks=undef
  Compiler:
    cc='cc', optimize='-O2', gccversion=egcs-2.91.66 19990314/Linux
    (egcs-1.1.2 release)
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include
    -D_LARGEFILE_SOURCE -D
_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
    lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:



@INC for perl v5.6.0:
    /usr/local/lib/perl5/5.6.0/i686-linux
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/jhoward
    LANG=en_US
    LANGUAGE (unset)
    LC_ALL=en_US
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/bin:/bin:/usr/sbin:/sbin:/usr/X11R6/bin:/usr/sbin:/usr/loca
    l/apache/bin:/root/bin:/usr/local/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

-- 
  Jeremy Howard
  jeremy@fastmail.fm


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

#### Created by jhoward@fastmail.fm



The following causes a 'Bizarre copy of HASH in aassign' error
perl \-MCGI&#8203;::Carp \-d \-e 'print croak\(1\)'
\<Then press 'c' to continue in the debugger>

It only occurs in debug mode\.

To fix it\, do the following&#8203;:
In Carp&#8203;::Heavy\, change 'DB' to 'DB\_tmp'&#8203;:
Line 39&#8203;: while \(do \{ \{ package DB\_tmp; @&#8203;a =3D caller\($i\+\+\) \} \} \) \{
Line 79&#8203;: @&#8203;a =3D @&#8203;\{DB\_tmp&#8203;::args\}; \# must get local copy of args

I don't know if there's any reason the DB package was used in
Carp&#8203;::Heavy\.\.\. If so\, this fix may break something\.

<details><summary>Perl Info</summary>

```
Flags:
    category=3Dlibrary
    severity=3Dhigh

Site configuration information for perl v5.6.0:

Configured by root at Fri May 19 12:49:37 EST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=3Dlinux, osvers=3D2.2.12-20, archname=3Di686-linux
    uname=3D'linux jlinux.optimal-decisions.com 2.2.12-20 #1 thu jan 6
    13:09:01 est 2000 i686 unknown '
    config_args=3D'-de'
    hint=3Drecommended, useposix=3Dtrue, d_sigaction=3Ddefine
    usethreads=3Dundef use5005threads=3Dundef useithreads=3Dundef
    usemultiplicity=3Dundef
    useperlio=3Dundef d_sfio=3Dundef uselargefiles=3Ddefine
    use64bitint=3Dundef use64bitall=3Dundef uselongdouble=3Dundef
    usesocks=3Dundef
  Compiler:
    cc=3D'cc', optimize=3D'-O2', gccversion=3Degcs-2.91.66 19990314/Linux
    (egcs-1.1.2 release)
    cppflags=3D'-fno-strict-aliasing -I/usr/local/include'
    ccflags =3D'-fno-strict-aliasing -I/usr/local/include
    -D_LARGEFILE_SOURCE -D
_FILE_OFFSET_BITS=3D64'
    stdchar=3D'char', d_stdstdio=3Ddefine, usevfork=3Dfalse
    intsize=3D4, longsize=3D4, ptrsize=3D4, doublesize=3D8
    d_longlong=3Ddefine, longlongsize=3D8, d_longdbl=3Ddefine, longdblsize=
=3D12
    ivtype=3D'long', ivsize=3D4, nvtype=3D'double', nvsize=3D8, Off_t=3D'of=
f_t',
    lseeksize=3D8
    alignbytes=3D4, usemymalloc=3Dn, prototype=3Ddefine
  Linker and Libraries:
    ld=3D'cc', ldflags =3D' -L/usr/local/lib'
    libpth=3D/usr/local/lib /lib /usr/lib
    libs=3D-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=3D/lib/libc-2.1.2.so, so=3Dso, useshrplib=3Dfalse, libperl=3Dlibpe=
rl.a
  Dynamic Linking:
    dlsrc=3Ddl_dlopen.xs, dlext=3Dso, d_dlsymun=3Dundef, ccdlflags=3D'-rdyn=
amic'
    cccdlflags=3D'-fpic', lddlflags=3D'-shared -L/usr/local/lib'

Locally applied patches:



@INC for perl v5.6.0:
    /usr/local/lib/perl5/5.6.0/i686-linux
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=3D/home/jhoward
    LANG=3Den_US
    LANGUAGE (unset)
    LC_ALL=3Den_US
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=3D/usr/bin:/bin:/usr/sbin:/sbin:/usr/X11R6/bin:/usr/sbin:/usr/loca
    l/apache/bin:/root/bin:/usr/local/bin
    PERL_BADLANG (unset)
    SHELL=3D/bin/bash

--=20
  Jeremy Howard
  jeremy@fastmail.fm


```
</details>

**Comment by p5pRT** [comment]
### From @gsar

On Wed\, 07 Jun 2000 16&#8203;:00&#8203;:57 \+0200\, Richard Foley wrote&#8203;:
>The following causes a 'Bizarre copy of HASH in aassign' error
>perl \-MCGI&#8203;::Carp \-d \-e 'print croak\(1\)'
>\<Then press 'c' to continue in the debugger>
>
>It only occurs in debug mode\.
\[\.\.\.\]
>I don't know if there's any reason the DB package was used in
>Carp&#8203;::Heavy\.\.\. If so\, this fix may break something\.

Yes\, that will break things\.

The real problem should be fixed in 5\.6\.1 \(coming sometime before
mid\-July\, I hope\)\.


Sarathy
gsar@&#8203;ActiveState\.com
